File: 4-median-of-two-sorted-arrays/median-of-two-sorted-arrays.java
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        int n1=nums1.length,n2=nums2.length;
        
        if(n2==0){
            if(n1%2==1) return nums1[n1/2];
            else return (nums1[n1/2]+nums1[(n1/2)-1])/2.0;
        }
        if(n1>n2) return findMedianSortedArrays(nums2,nums1);
        int l=0,h=n1,cut1=0,cut2=0,l1=0,l2=0,h1=0,h2=0,mid;
        
        cut1=n1;
        while(l<=h){
           cut1=(l+h)>>1;
           cut2=((n1+n2+1)/2)-cut1;

           if(cut1==0) l1=-(int)1e9;
           else l1=nums1[cut1-1];
           if(cut2==0) l2=-(int)(1e9);
           else l2=nums2[cut2-1];

           if(cut1==n1) h1=(int)1e9;
           else h1=nums1[cut1];
           if(cut2==n2) h2=(int)1e9;
           else h2=nums2[cut2];

           
           if(l1>h2) h=cut1-1;
           else if(l2>h1) l=cut1+1;
           else{
               if((n1+n2)%2==0){
                   return ((Math.max(l1,l2)+Math.min(h1,h2))/2.0);
               }
               else return Math.max(l1,l2);
           }
        }

        return 0;
    }
}

File: 4-median-of-two-sorted-arrays/median-of-two-sorted-arrays.py
class Solution(object):
    def findMedianSortedArrays(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: float
        """
        n1 = len(nums1)
        n2 = len(nums2)
        if(n1 > n2) : return self.findMedianSortedArrays(nums2, nums1)

        n = n1 + n2 
        left = (n1 + n2 + 1) / 2
        l = 0
        h = n1 

        while(l <= h):
            mid = (l + h) / 2
            mid2 = left - mid

            l1 = -(10 ** 10)
            l2 = -(10 ** 10)
            r1 = (10 ** 10)
            r2 = (10 ** 10)

            if(mid > 0) : l1 = nums1[mid - 1]
            if(mid2 > 0): l2 = nums2[mid2 - 1]

            if(mid < n1): r1 = nums1[mid]
            if(mid2 < n2): r2 = nums2[mid2]

            if(max(l1, l2) <= min(r1, r2)):
                if((n1 + n2) % 2 == 0):
                    return (float((max(l1, l2)) + float(min(r1, r2))) / 2.0)
                else:
                    return max(l1, l2)
            
            if(l1 > l2 and l1 > min(r1, r2)):
                h = mid - 1
            elif(l2>l1 and l2 > min(r1, r2)):
                l = mid + 1
        
        return 0.0



        